### ISSUE TITLE
[Windows] ImportError: Unable to locate juce library path

### ISSUE THREAD

**Issue by kokounet** [kunitoki/popsicle#3]
Hello! I'm trying to use `popsicle` on Windows but I encounter the following error when running the simple example from the readme:

```
ImportError: Unable to locate juce library path
```

Having debugged a little it seems that in the function `__get_juce_library_path` the following variables are not pointing at the right directories:

```
base_path = [...]\site-packages\popsicle\..\cxxbuild\popsicle_artefacts
library_path = [...]\site-packages\cxxbuild\popsicle_artefacts\libpopsicle.dll
```

but on windows the `cxxbuild` directory does not exist in `site-packages` plus the dll is called `popsicle.dll` and is located at `[...]\site-packages\popsicle\data\` (as well as all the juce headers)

**Comment by kunitoki** [comment]
Are you compiling locally or using the wheel in pypi ? i will have a look, current version in master should be improved in this regard but is not published yet

**Comment by kunitoki** [comment]
can you try installing 0.0.7 ? https://pypi.org/project/popsicle/0.0.7/#files

**Comment by kokounet** [comment]
Hello!
I am compiling locally as there is no wheel available for cppyy if I understood correctly.
Using `0.0.7` I get this output instead:
```
(Re-)building pre-compiled headers (options: -O2); this may take a minute ...
Traceback (most recent call last):
  File "C:\Users\Chris\devel\playground\pyjuce\main.py", line 1, in <module>
    from popsicle import juce_gui_basics
  File "C:\Users\Chris\devel\playground\pyjuce\.venv\lib\site-packages\popsicle\__init__.py", line 7, in <module>
    from . import juce_events as __juce_events
  File "C:\Users\Chris\devel\playground\pyjuce\.venv\lib\site-packages\popsicle\juce_events.py", line 6, in <module>
    from . import juce_core as __juce_core
  File "C:\Users\Chris\devel\playground\pyjuce\.venv\lib\site-packages\popsicle\juce_core.py", line 27, in <module>
    __juce_bootstrap()
  File "C:\Users\Chris\devel\playground\pyjuce\.venv\lib\site-packages\popsicle\utils.py", line 96, in juce_bootstrap
    __cppyy.load_library(__get_juce_library_path(root_path))
  File "C:\Users\Chris\devel\playground\pyjuce\.venv\lib\site-packages\popsicle\utils.py", line 35, in __get_juce_library_path
    if target: base_path = __os.path.join(base_path, target)
  File "C:\Program Files\Python39\lib\ntpath.py", line 78, in join
    path = os.fspath(path)
TypeError: expected str, bytes or os.PathLike object, not list
```

This line in `utils.py:33` returns an empty list on my machine:
```py
base_path = __glob.glob(__os.path.join(root_path, "..", "build", "**", "popsicle_artefacts"))
```

I hope this helps,
Cheers!

**Comment by kunitoki** [comment]
well there should be a version in pypi for windows, but it's not fully tested. try running `pip install popsicle`.

how did you install it into the virtual environment ? did you build the wheel locally ? can you list here the content of  the `C:\Users\Chris\devel\playground\pyjuce\.venv\lib\site-packages\popsicle` folder and subfolders ?

**Comment by kunitoki** [comment]
I've pushed the fix for the dll name. You should be able to compile it locally now

**Comment by kokounet** [comment]
So my process was indeed to do a `pip install popsicle` in my project and copying the example code from the readme to `main.py` and just running it. I suspected I recompiled the thing locally because of this message from `cppyy`, but I might be wrong:
```
(Re-)building pre-compiled headers (options: -O2); this may take a minute ...
```
Here's the tree for the popsicle folder:

```
│   juce_analytics.py
│   juce_audio_basics.py
│   juce_audio_devices.py
│   juce_audio_formats.py
│   juce_audio_processors.py
│   juce_audio_utils.py
│   juce_core.py
│   juce_cryptography.py
│   juce_data_structures.py
│   juce_dsp.py
│   juce_events.py
│   juce_graphics.py
│   juce_gui_basics.py
│   juce_gui_extra.py
│   juce_opengl.py
│   juce_osc.py
│   juce_video.py
│   utils.py
│   __init__.py
│
└───data
    │   Defs.txt
    │   popsicle.dll
    │
    │ (I omit the files in those folder, they seem ok)
    ├───juce_analytics 
    ├───juce_audio_basics
    ├───juce_audio_devices
    ├───juce_audio_formats
    ├───juce_audio_plugin_client
    ├───juce_audio_processors
    ├───juce_audio_utils
    ├───juce_blocks_basics
    ├───juce_box2d
    ├───juce_core
    ├───juce_cryptography
    ├───juce_data_structures
    ├───juce_dsp
    ├───juce_events
    ├───juce_graphics
    ├───juce_gui_basics
    ├───juce_gui_extra
    ├───juce_opengl
    ├───juce_osc
    ├───juce_product_unlocking
    └───juce_video
```

I'll try to clone an insall `popsicle` locally in the same venv see if it works now 😅 .

**Comment by kunitoki** [comment]
can you try `pip install -I popsicle==0.0.8` ?

**Comment by kokounet** [comment]
So I built popsicle locally and running the example `animated_component.py` and now I get this error:
```
(.venv) C:\Users\Chris\devel\libs\popsicle>python examples\animated_component.py
(Re-)building pre-compiled headers (options: -O2); this may take a minute ...
Traceback (most recent call last):
  File "C:\Users\Chris\devel\libs\popsicle\examples\animated_component.py", line 6, in <module>
    from popsicle import juce_gui_basics, juce_gui_extra
  File "C:\Users\Chris\devel\libs\popsicle\.venv\lib\site-packages\popsicle\__init__.py", line 7, in <module>
    from . import juce_events as __juce_events
  File "C:\Users\Chris\devel\libs\popsicle\.venv\lib\site-packages\popsicle\juce_events.py", line 6, in <module>
    from . import juce_core as __juce_core
  File "C:\Users\Chris\devel\libs\popsicle\.venv\lib\site-packages\popsicle\juce_core.py", line 28, in <module>
    __juce_include()
  File "C:\Users\Chris\devel\libs\popsicle\.venv\lib\site-packages\popsicle\juce_core.py", line 11, in __juce_include
    __cppyy.include("juce_core/juce_core.h")
  File "C:\Users\Chris\devel\libs\popsicle\.venv\lib\site-packages\cppyy\__init__.py", line 229, in include
    raise ImportError('Failed to load header file "%s"%s' % (header, err.err))
ImportError: Failed to load header file "juce_core/juce_core.h"
input_line_23:1:10: fatal error: 'juce_core/juce_core.h' file not found
#include "juce_core/juce_core.h"
         ^~~~~~~~~~~~~~~~~~~~~~~
```

Same goes when I `pip install -I popsicle==0.0.8` on my example project: 
```
(.venv) C:\Users\Chris\devel\playground\pyjuce>python main.py
(Re-)building pre-compiled headers (options: -O2); this may take a minute ...
Traceback (most recent call last):
  File "C:\Users\Chris\devel\playground\pyjuce\main.py", line 1, in <module>
    from popsicle import juce_gui_basics
  File "C:\Users\Chris\devel\playground\pyjuce\.venv\lib\site-packages\popsicle\__init__.py", line 7, in <module>
    from . import juce_events as __juce_events
  File "C:\Users\Chris\devel\playground\pyjuce\.venv\lib\site-packages\popsicle\juce_events.py", line 6, in <module>
    from . import juce_core as __juce_core
  File "C:\Users\Chris\devel\playground\pyjuce\.venv\lib\site-packages\popsicle\juce_core.py", line 28, in <module>
    __juce_include()
  File "C:\Users\Chris\devel\playground\pyjuce\.venv\lib\site-packages\popsicle\juce_core.py", line 11, in __juce_include
    __cppyy.include("juce_core/juce_core.h")
  File "C:\Users\Chris\devel\playground\pyjuce\.venv\lib\site-packages\cppyy\__init__.py", line 229, in include
    raise ImportError('Failed to load header file "%s"%s' % (header, err.err))
ImportError: Failed to load header file "juce_core/juce_core.h"
input_line_23:1:10: fatal error: 'juce_core/juce_core.h' file not found
#include "juce_core/juce_core.h"
         ^~~~~~~~~~~~~~~~~~~~~~~
```

I hope this helps! 🙏

**Comment by kunitoki** [comment]
It seems the include paths are not setup correctly

**Comment by jellyr** [comment]
> So I built popsicle locally and running the example `animated_component.py` and now I get this error:
> 
> ```
> (.venv) C:\Users\Chris\devel\libs\popsicle>python examples\animated_component.py
> (Re-)building pre-compiled headers (options: -O2); this may take a minute ...
> Traceback (most recent call last):
>   File "C:\Users\Chris\devel\libs\popsicle\examples\animated_component.py", line 6, in <module>
>     from popsicle import juce_gui_basics, juce_gui_extra
>   File "C:\Users\Chris\devel\libs\popsicle\.venv\lib\site-packages\popsicle\__init__.py", line 7, in <module>
>     from . import juce_events as __juce_events
>   File "C:\Users\Chris\devel\libs\popsicle\.venv\lib\site-packages\popsicle\juce_events.py", line 6, in <module>
>     from . import juce_core as __juce_core
>   File "C:\Users\Chris\devel\libs\popsicle\.venv\lib\site-packages\popsicle\juce_core.py", line 28, in <module>
>     __juce_include()
>   File "C:\Users\Chris\devel\libs\popsicle\.venv\lib\site-packages\popsicle\juce_core.py", line 11, in __juce_include
>     __cppyy.include("juce_core/juce_core.h")
>   File "C:\Users\Chris\devel\libs\popsicle\.venv\lib\site-packages\cppyy\__init__.py", line 229, in include
>     raise ImportError('Failed to load header file "%s"%s' % (header, err.err))
> ImportError: Failed to load header file "juce_core/juce_core.h"
> input_line_23:1:10: fatal error: 'juce_core/juce_core.h' file not found
> #include "juce_core/juce_core.h"
>          ^~~~~~~~~~~~~~~~~~~~~~~
> ```
> 
> Same goes when I `pip install -I popsicle==0.0.8` on my example project:
> 
> ```
> (.venv) C:\Users\Chris\devel\playground\pyjuce>python main.py
> (Re-)building pre-compiled headers (options: -O2); this may take a minute ...
> Traceback (most recent call last):
>   File "C:\Users\Chris\devel\playground\pyjuce\main.py", line 1, in <module>
>     from popsicle import juce_gui_basics
>   File "C:\Users\Chris\devel\playground\pyjuce\.venv\lib\site-packages\popsicle\__init__.py", line 7, in <module>
>     from . import juce_events as __juce_events
>   File "C:\Users\Chris\devel\playground\pyjuce\.venv\lib\site-packages\popsicle\juce_events.py", line 6, in <module>
>     from . import juce_core as __juce_core
>   File "C:\Users\Chris\devel\playground\pyjuce\.venv\lib\site-packages\popsicle\juce_core.py", line 28, in <module>
>     __juce_include()
>   File "C:\Users\Chris\devel\playground\pyjuce\.venv\lib\site-packages\popsicle\juce_core.py", line 11, in __juce_include
>     __cppyy.include("juce_core/juce_core.h")
>   File "C:\Users\Chris\devel\playground\pyjuce\.venv\lib\site-packages\cppyy\__init__.py", line 229, in include
>     raise ImportError('Failed to load header file "%s"%s' % (header, err.err))
> ImportError: Failed to load header file "juce_core/juce_core.h"
> input_line_23:1:10: fatal error: 'juce_core/juce_core.h' file not found
> #include "juce_core/juce_core.h"
>          ^~~~~~~~~~~~~~~~~~~~~~~
> ```
> 
> I hope this helps! 🙏

just delete all the juce modules in site-packages\popsicle\data  and  recopy these modules from  "https://github.com/kunitoki/JUCE/tree/99112cf71f0cc3aafa4622b62cf81f2b7df38f0b/modules"  , 
then errors like fatal error: 'juce_core/juce_core.h' file not found  will be gone!
bcz the original files in juce module folder is not right.

**Comment by kunitoki** [comment]
I've pushed 0.0.9 that should work better in this regard. In case you still have issues, please reopen a ticket.

**Comment by e1pupper** [comment]
Hey @kunitoki i recently did pip install popsicle and tried to run the example you provided but i got the same error as them not sure what to do
    raise ImportError('Failed to load header file "%s"%s' % (header, err.err))
ImportError: Failed to load header file "juce_core/juce_core.h"
input_line_23:1:10: fatal error: 'juce_core/juce_core.h' file not found
#include "juce_core/juce_core.h"
         ^~~~~~~~~~~~~~~~~~~~~~~

**Comment by kunitoki** [comment]
I can confirm that windows wheels are broken, i already merged a fix and will release the next version soon

**Comment by e1pupper** [comment]
> I can confirm that windows wheels are broken, i already merged a fix and will release the next version soon

appreciatge it <3

**Comment by kunitoki** [comment]
Should be fixed with 0.0.10, pip upgrade it again.

In case you have more issues, reopen a new ticket @e1pupper
